### PR TITLE
My Jetpack: Add 'unlinked' parameter to the checkout URL when the user is not connected

### PR DESCRIPTION
**Pull request by retrofox**
Fixes https://github.com/Automattic/jetpack/issues/22731

#### Changes proposed in this Pull Request:

* My Jetpack: Adds `unlinked` parameter to the checkout URL when the user is not connected

#### Testing instructions:

* Go to My Jetpack
* Ensure the user is not connected to Jetpack
* Go to the add-search route
* Confirm, once the user clicks on the Add button, before to gets the Checkout page, it will try to connect with Jetpack.
`https://<your-testing-site>/wp-admin/admin.php?page=my-jetpack#/add-search`

https://user-images.githubusercontent.com/77539/153032585-fab310a0-9cc9-4440-aa9e-f7e015ef2c5f.mov

### CODE REVIEW

**Comment by leogermani** [comment]
It seems that the `redirect_to` param, with a query string of its own, is breaking the URL.

The `unlinked` param was added but it was not processed by Calypso. I was taken directly to the checkout page.

If I removed the `?` and `&` from the `redicrect_to`, then I was correctly taken to connect my user before the checkout. And the user was correctly added to the site afterwards.

We need to figure out how to deal with this when building the URL

**Comment by retrofox** [comment]
Addressed [here](https://github.com/Automattic/jetpack/pull/22740/commits/c7e71af707ccea2386730ab480221f4d3d0be30b). though.


https://user-images.githubusercontent.com/77539/153093822-67307d20-057b-4394-9f6c-8b08ffeb8cee.mov

File: projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js [hunk 1]
```diff
@@ -2,29 +2,20 @@
...
-		site,
-		path: 'jetpack_search',
-		query: `redirect_to=${ redirect_to }`,
-	} );
+export default function getProductCheckoutUrl( product, isUserConnected ) {
+	const { siteSuffix, redirectUrl } = window?.myJetpackInitialState || {};
+	const redirect_to = `${ redirectUrl.replace( /=/g, '%3d' ) }${
+		! isUserConnected ? '&unlinked=1' : ''
+	}&site=${ siteSuffix }`;
+	return `https://wordpress.com/checkout/${ siteSuffix }/${ product }?redirect_to=${ redirect_to }`;
```
Comment (renatoagds) [inline projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js:20]: Did you try to use the `URL` utility from the browser to handle all of this?

https://developer.mozilla.org/en-US/docs/Web/API/URL/URL

It provides good handling of `searchParams`:

https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams

File: projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js [hunk 2]
```diff
@@ -2,29 +2,20 @@
...
-	const redirect_to = `${ redirectUrl }&product=${ product }`;
-
-	return getRedirectUrl( MY_JETPACK_PRODUCT_CHECKOUT, {
-		site,
-		path: 'jetpack_search',
-		query: `redirect_to=${ redirect_to }`,
-	} );
+export default function getProductCheckoutUrl( product, isUserConnected ) {
+	const { siteSuffix, redirectUrl } = window?.myJetpackInitialState || {};
+	const redirect_to = `${ redirectUrl.replace( /=/g, '%3d' ) }${
```
Comment (renatoagds) [inline projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js:17]: You could use `encodeURIComponent` to handle the `=` to `%3d`, not?

https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent

Comment (retrofox) [reply projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js:17]: Yes, I think we can use it partially. we can't encode the whole URL. Let me use it at least for the My Jetpack page.

File: projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js (hunk 1 above)
Comment (retrofox) [reply projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js:20]: It doesn't seem to be completely supported by all browsers. Looking...

File: projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js (hunk 2 above)
Comment (retrofox) [reply projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js]: We can encode only the `redirectUrl`. This is something that maybe we can address in the Jetpack Redirect system. Addressed [here](https://github.com/Automattic/jetpack/pull/22740/commits/b7c6debf6f3ef4cc2e64a246964909bc526a393c).

File: projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js (hunk 1 above)
Comment (retrofox) [reply projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js:20]: > It provides good handling of searchParams:

> https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams

Is it to build or to stripe them? According to the doc:
> The searchParams readonly property of the [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) interface returns a [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) object allowing access to the [GET](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET) decoded query arguments contained in the URL.

Comment (renatoagds) [reply projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js:20]: Just linking back to our discussion in slack 🧵  about Safari support and write access: p1644363459463389-slack-C02TQF5VAJD

Comment (retrofox) [reply projects/packages/my-jetpack/_inc/utils/get-product-checkout-url.js]: Addressed [here](https://github.com/Automattic/jetpack/pull/22740/commits/56265321e2d677934436c6d2ad19f8a7db6132fe).

**Review by renatoagds — APPROVED** [review]

**Pull request merged**